gcd/src/main/java/org/spine3/server/storage/datastore/DsEventStorage.java

```diff
@@ -20,100 +20,293 @@
             return null;
         }
 
-        final EntityResult entity = response.getFound(0);
-        final EventStorageRecord result = entityToMessage(entity, TYPE_URL);
+        final EventStorageRecord result = entityToMessage(response, RECORD_TYPE_URL);
         return result;
     }
 
-    private static final Function<EntityResult, Event> ENTITY_TO_EVENT = new Function<EntityResult, Event>() {
-        @Nullable
+    private static class EventPredicate implements Predicate<EventStorageRecord> {
+
+        private final Collection<EventFilter> eventFilters;
+
+        private EventPredicate(Collection<EventFilter> eventFilters) {
+            this.eventFilters = eventFilters;
+        }
+
+        @SuppressWarnings("MethodWithMoreThanThreeNegations")
         @Override
-        public Event apply(@Nullable EntityResult entityResult) {
-            if (entityResult == null) {
-                return Event.getDefaultInstance();
+        public boolean apply(@Nullable EventStorageRecord event) {
+            if (event == null) {
+                return false;
+            }
+
+            if (eventFilters.isEmpty()) {
+                return true;
             }
-            final EventStorageRecord message = entityToMessage(entityResult, TYPE_URL);
-            final Event result = toEvent(message);
+
+            for (EventFilter filter : eventFilters) {
+                final EventFilterChecker predicate = new EventFilterChecker(filter);
+                if (predicate.isDefault()) {
+                    continue;
+                }
+                final boolean matches = predicate.apply(event);
+                if (matches) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    private static class EventFilterChecker implements Predicate<EventStorageRecord> {
+
+        private final String eventType;
+        private final Collection<String> aggregateIds;
+        private final Collection<FieldFilter> eventFieldFilters;
+        private final Collection<FieldFilter> contextFieldFilters;
+
+        private static final Function<Any, Message> ANY_UNPACKER = new Function<Any, Message>() {
+            @Nullable
+            @Override
+            public Message apply(@Nullable Any input) {
+                if (input == null) {
+                    return null;
+                }
+
+                return AnyPacker.unpack(input);
+            }
+        };
+
+        private EventFilterChecker(@SuppressWarnings("TypeMayBeWeakened") EventFilter eventFilter) {
+            this.eventType = checkNotNull(eventFilter.getEventType());
+            this.aggregateIds = Collections2.transform(eventFilter.getAggregateIdList(), ID_TRANSFORMER);
+            this.eventFieldFilters = eventFilter.getEventFieldFilterList();
+            this.contextFieldFilters = eventFilter.getContextFieldFilterList();
+        }
+
+        private boolean isDefault() {
```

Comment:
By analogy with protobuf types (`Message::getDefaultInstance`) default is the state of the object when all fields are empty.